### PR TITLE
added lock option for enqueue and retry methods (#7)

### PR DESCRIPTION
**Pull request by avkhozov**

### CODE REVIEW

**Comment by kraih** [comment]
The last time this feature was discussed it did not have a lot of support.

**Comment by avkhozov** [comment]
Yes, I see. But I think ```minion``` not yet such popular as ```mojolicious``` for example, and it now has a small feedback.

**Comment by kraih** [comment]
Well, personally i have no need for this feature, so maybe i should start with a :-1: vote. Especially since i've not seen anything similar in other job queues yet, so this is not a proven approach.

**Comment by jberger** [comment]
What is the performance impact when this feature is not used?

**Comment by avkhozov** [comment]
I have already described example of usage in https://github.com/kraih/minion/issues/7#issuecomment-167209845

For example, you have two servers with the same configuration (mojo app + minion worker) and you need run periodically jobs (enqueued via cron). So you must solve the problem with duplication of same tasks. Without ``lock`` option in ``minion`` you need to manage this manually by creating extra tables/collections in your app and write logic for deduplication and cleaning old lock keys, etc. In minion backends this problems solved easier and minion more suited for this.

Of course, this is only my opinion.

File: lib/Minion/Backend/Pg.pm [hunk 1]
```diff
@@ -28,14 +28,17 @@ sub enqueue {
...
 
-  my $db = $self->pg->db;
-  return $db->query(
-    "insert into minion_jobs (args, attempts, delayed, priority, queue, task)
-     values (?, ?, (now() + (interval '1 second' * ?)), ?, ?, ?)
-     returning id", {json => $args}, $options->{attempts} // 1,
-    $options->{delay} // 0, $options->{priority} // 0,
-    $options->{queue} // 'default', $task
-  )->hash->{id};
+  my $result = eval {
```
Comment (kraih) [inline lib/Minion/Backend/Pg.pm]: Wouldn't an `ON CONFLICT DO NOTHING` work much better than `eval`?

**Comment by kraih** [comment]
Regarding what features to accept, as long as we don't have more users willing to vote, i think features will be limited to what i need and/or what most other job queues support.

File: lib/Minion/Backend/Pg.pm (hunk 1 above)
Comment (avkhozov) [inline lib/Minion/Backend/Pg.pm]: Yes, here it would be appropriate. But it requires a postgres 9.5.

Comment (kraih) [inline lib/Minion/Backend/Pg.pm]: Minion will soon require PostgreSQL 9.5 anyway for `SKIP LOCKED`.

Comment (avkhozov) [inline lib/Minion/Backend/Pg.pm]: Ok, I'll fix this when config for travis will be improved.

**Comment by avkhozov** [comment]
Also there are recurring/periodic jobs (https://github.com/mperham/sidekiq/wiki/Ent-Periodic-Jobs) that cover my needs with lock option. I think it's more general case, which can enjoy more popularity.

**Comment by kraih** [comment]
Yes, periodic jobs does look like a pretty nice feature, but also really hard to get right, makes sense Sidekiq only has it in the commercial version.

**Comment by kraih** [comment]
But back to what this pull request is about (unique jobs), why does the `lock` option use an arbitrary key at all? Why not just a boolean column called? The name `lock` is also not particularly nice, `unique => 1` would look much more appropriate.

**Comment by kraih** [comment]
Unique jobs in sidekiq also have a time period attached. https://github.com/mperham/sidekiq/wiki/Ent-Unique-Jobs

**Comment by avkhozov** [comment]
Hmm, I thought that the user should care about what jobs need not be repeated specifying a unique key for them. How ``minion`` will know that the job should not be enqueued if the user just specify ``unique => 1``?

The decision of sidekiq "1 job for a 1 task in a certain period" also looks like a working.

**Comment by kraih** [comment]
Looks like we can't reach consensus on what the right way for implementing unique jobs would be. The question about performance cost is also still unanswered. And it appears that what everyone asking for unique jobs really wants is periodic jobs, perhaps we should consider that first.

**Comment by kraih** [comment]
I think a feature like this is still very likely to be added in the future, but we have to look at it from all angles first, the real goal is scheduled/periodic jobs.

**Comment by kraih** [comment]
Just so i don't forget to mention it in future discussions, the way the `lock` value is handled in `retry_job` is very sketchy.

**Pull request closed without merging**